p2p: Allow block downloads from peers without snapshot block after assumeutxo validation

stringintech opened the pull request:
Currently, after assumeutxo background validation finishes, the node continues to skip peers that don't have the snapshot block in their best chain until restart. This unnecessarily excludes peers from block downloads even though the background sync has completed and undo data is available.

The restriction persists because `m_chainman.CurrentChainstate().SnapshotBase()` continues to return the snapshot base block until restart, even after validation completes. Added `m_chainman.CurrentChainstate().m_assumeutxo == Assumeutxo::UNVALIDATED` check to only apply the peer restriction while background validation is ongoing. 

Also added test coverage in `feature_assumeutxo.py` that verifies peers without the snapshot block can be used for block downloads after background validation completes. The test fails without this fix.

stratospher reviewed (commented):
ACK 9f946a7.

nice! I think this aligns with the original intention to restrict the peer only when background sync isn't over in https://github.com/bitcoin/bitcoin/pull/29519#discussion_r1635456871 where this code was introduced.

looks like we might incorrectly set `pindexLastCommonBlock` in [L1401](https://github.com/bitcoin/bitcoin/blob/9f946a79ca894741547c5892f9dc60d023ae1508/src/net_processing.cpp#L1401) to the previous best known block tip again in this scenario. but it's not a problem since we immediately reset it back to the correct value + comments suggest it's ok to set it incorrectly and that's way better than introducing more conditions.

also slightly related: https://github.com/bitcoin/bitcoin/issues/30288

mzumsande reviewed (commented):
Concept ACK

fjahr commented on src/net_processing.cpp:
```diff
@@ -1388,7 +1388,7 @@ void PeerManagerImpl::FindNextBlocksToDownload(const Peer& peer, unsigned int co
     // We can't reorg to this chain due to missing undo data until the background sync has finished,
```
nit: Could clarify the behavior in the comment here a bit more.

stringintech committed:
p2p: Allow block downloads from peers without snapshot block after assumeutxo validation

After assumeutxo background validation completes, allow block downloads from peers that don't have the snapshot block in their best chain.

Previously, these peers were skipped until restart because `m_chainman.CurrentChainstate().SnapshotBase()` continued returning non-null even after validation finished. Add `m_chainman.CurrentChainstate().m_assumeutxo == Assumeutxo::UNVALIDATED` check to only apply the restriction while background validation is ongoing.

stringintech committed:
test: Verify peer usage after assumeutxo validation completes

Add test coverage to ensure peers without the snapshot block in their chain can be used for block downloads after background validation completes. The test fails without the fix in the previous commit.

stringintech commented:
9f946a7 to 6b92a69 - rebased, resolved conflict with #30214, and addressed https://github.com/bitcoin/bitcoin/pull/33604#discussion_r2628258598. Also PR and first commit descriptions are updated.

fjahr reviewed (commented):
tACK 5cd096803d223166b9aa36e783406a36cdb150bc

Reviewed code and confirmed the new test failed previous to the behavior change.

sedited reviewed (approved):
ACK 5cd096803d223166b9aa36e783406a36cdb150bc

sedited reviewed (approved):
Re-ACK 7d9e1a810239a65a153c35f0f94490560441db49

achow101 commented:
ACK 7d9e1a810239a65a153c35f0f94490560441db49
